RadarSDK/RadarLocationPremissionsStatus.m

```diff
@@ -0,0 +1,154 @@
+            break;
+        case kCLAuthorizationStatusAuthorizedWhenInUse:
+            statusString = @"AuthorizedWhenInUse";
+            break;
+        default:
+            statusString = @"Unknown";
+    }
+    return @{
+        @"locationManagerStatus": statusString,
+        @"backgroundPopupAvailable": @(self.backgroundPopupAvailable),
+        @"inForegroundPopup": @(self.inForegroundPopup),
+        @"userRejectedBackgroundPermissions": @(self.userRejectedBackgroundPermissions),
+        @"locationPermissionState": [RadarLocationPermissionsStatus stringForLocationPermissionState:self.locationPermissionState]
+    };
+}
+
+- (instancetype _Nullable)initWithDictionary:(NSDictionary *)dictionary {
+    NSString *statusString = dictionary[@"locationManagerStatus"];
+    CLAuthorizationStatus locationManagerStatus;
+    if ([statusString isEqualToString:@"NotDetermined"]) {
+        locationManagerStatus = kCLAuthorizationStatusNotDetermined;
+    } else if ([statusString isEqualToString:@"Restricted"]) {
+        locationManagerStatus = kCLAuthorizationStatusRestricted;
+    } else if ([statusString isEqualToString:@"Denied"]) {
+        locationManagerStatus = kCLAuthorizationStatusDenied;
+    } else if ([statusString isEqualToString:@"AuthorizedAlways"]) {
+        locationManagerStatus = kCLAuthorizationStatusAuthorizedAlways;
+    } else if ([statusString isEqualToString:@"AuthorizedWhenInUse"]) {
+        locationManagerStatus = kCLAuthorizationStatusAuthorizedWhenInUse;
+    } else {
+        locationManagerStatus = kCLAuthorizationStatusNotDetermined;
+    }
+    BOOL backgroundPopupAvailable = [dictionary[@"backgroundPopupAvailable"] boolValue];
+    BOOL inForegroundPopup = [dictionary[@"inForegroundPopup"] boolValue];
+    BOOL userRejectedBackgroundPermissions = [dictionary[@"userRejectedBackgroundPermissions"] boolValue];
+    return [self initWithStatus:locationManagerStatus 
+       backgroundPopupAvailable:backgroundPopupAvailable 
+              inForegroundPopup:inForegroundPopup 
+userRejectedBackgroundPermissions:userRejectedBackgroundPermissions];
+}
+
++ (NSString *)stringForLocationPermissionState:(RadarLocationPermissionState)state {
+    switch (state) {
+        case NoPermissionsGranted:
+            return @"NoPermissionsGranted";
+        case ForegroundPermissionsGranted:
+            return @"ForegroundPermissionsGranted";
+        case ForegroundPermissionsRejected:
+            return @"ForegroundPermissionsRejected";
+        case ForegroundPermissionsPending:
+            return @"ForegroundPermissionsPending";
+        case BackgroundPermissionsGranted:
+            return @"BackgroundPermissionsGranted";
+        case BackgroundPermissionsRejected:
+            return @"BackgroundPermissionsRejected";
+        case BackgroundPermissionsPending:
+            return @"BackgroundPermissionsPending";
+        case PermissionsRestricted:
+            return @"PermissionsRestricted";
```

Comment:
do we have the "allow once" case covered? by which?

Reply:
So the `allow once` case has 2 "phases". On the first session, it is functionally similar to the `granted foreground` state, with the exception that we cannot launch the request background pop-up (which is handled by us detecting that no pop-up is launched and sending user to recovery state).  On the subsequent session, it is similar to `no permissions granted` and we handle it simply by detecting changes in permissions.